### PR TITLE
Added use-case testing support to upgrade testing.

### PR DESCRIPTION
**Pull request by san7ket**
Adding use cases tests, to be tested with the upgrade process.
This tends , to capture corner cases , that our traditional/functional testing tends to miss out.
This will ensure to configure entities/features/contents(CV's)/provision hosts , before upgrade do some transactions on it , aslo asserts this pre-upgarde.
After upgrade, we are able to interact with same entities/ features/hosts created pre-upgrade , perform sanity checks, add corner cases to those entities.
issue: https://github.com/SatelliteQE/automation-tools/issues/532

Refactored PR from https://github.com/SatelliteQE/automation-tools/pull/569

### CODE REVIEW

File: upgrade_tests/helpers/scenarios_utils.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    get_attribute_value,
+    hammer,
+)
+from upgrade.helpers.docker import remove_all_docker_containers
+from upgrade.helpers.rhevm import (
+    create_rhevm_instance,
+    get_rhevm_client,
+    wait_till_rhevm_instance_status
+)
+from automation_tools.satellite6.upgrade.tools import logger
```
Comment (jyejare) [inline upgrade_tests/helpers/scenarios_utils.py]: You have logger in upgrades itself. Import from there, logger from automation tools will be deleted soon.

File: upgrade_tests/helpers/scenarios_utils.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
+"""All the helper functions, needed for scenarios test case automation to be
+added here"""
+import os
+import pickle
```
Comment (jyejare) [inline upgrade_tests/helpers/scenarios_utils.py]: Use json instead.

File: upgrade_tests/helpers/scenarios_utils.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+data = {}
+logger = logger()
+
+
+def create_dict(entities_dict):
+    """Stores a dictionary of entities created in satellite on the disk
+
+        :param string entities_dict: A dictionary of entities created in
+        satellite
+        """
```
Comment (jyejare) [inline upgrade_tests/helpers/scenarios_utils.py]: Indentation issue.

File: upgrade_tests/helpers/scenarios_utils.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+rpm1 = 'https://inecas.fedorapeople.org/fakerepos/zoo3/bear-4.1-1.noarch.rpm'
+rpm2 = 'https://inecas.fedorapeople.org/fakerepos/zoo3/camel-0.1-1.noarch.rpm'
+data = {}
+logger = logger()
+
+
+def create_dict(entities_dict):
+    """Stores a dictionary of entities created in satellite on the disk
+
+        :param string entities_dict: A dictionary of entities created in
```
Comment (jyejare) [inline upgrade_tests/helpers/scenarios_utils.py]: Indentation issue.

File: upgrade_tests/helpers/scenarios_utils.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        satellite
+        """
+    data.update(entities_dict)
+    with open('entities_data', 'wb') as pref:
```
Comment (jyejare) [inline upgrade_tests/helpers/scenarios_utils.py]: I think we should go away from creating so many files of dicts, we may create a single file with all the data. Or at least should categories these files.

File: upgrade_tests/helpers/scenarios_utils.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+from fabric.api import execute, run
+
+rpm1 = 'https://inecas.fedorapeople.org/fakerepos/zoo3/bear-4.1-1.noarch.rpm'
+rpm2 = 'https://inecas.fedorapeople.org/fakerepos/zoo3/camel-0.1-1.noarch.rpm'
+data = {}
+logger = logger()
+
+
+def create_dict(entities_dict):
+    """Stores a dictionary of entities created in satellite on the disk
```
Comment (jyejare) [inline upgrade_tests/helpers/scenarios_utils.py]: Need more explanation wrt to scenario tests.

File: upgrade_tests/helpers/scenarios_utils.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        :param string entities_dict: A dictionary of entities created in
+        satellite
+        """
+    data.update(entities_dict)
+    with open('entities_data', 'wb') as pref:
+        pickle.dump(data, pref)
+
+
+def get_entity_data(scenario_name):
+    """Fetches the dictionary of entities from the disk depending on the
```
Comment (jyejare) [inline upgrade_tests/helpers/scenarios_utils.py]: Need more explanation wrt to scenario tests.

File: upgrade_tests/helpers/scenarios_utils.py [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+    Scenario name (class name in which test is defined)
+
+    :param string scenario_name : The name of the class for which the data is
+    to fetched
+
+    :returns dict entity_data : Returns a dictionary of entities
+    """
+    with open('entities_data') as pref:
+        entity_data = pickle.load(pref)
```
Comment (jyejare) [inline upgrade_tests/helpers/scenarios_utils.py]: use json instead.

File: upgrade_tests/helpers/scenarios_utils.py [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+        entity_data = entity_data[scenario_name]
+    return entity_data
+
+
+def generate_satellite_docker_clients_on_rhevm_upgrade(
```
Comment (jyejare) [inline upgrade_tests/helpers/scenarios_utils.py]: Why duplication of code https://github.com/SatelliteQE/satellite6-upgrade/blob/master/upgrade/helpers/docker.py#L11.

You can modify existing to match ur need.

File: upgrade_tests/helpers/scenarios_utils.py [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+                'docker run -d -h {0} -e "SATHOST={1}" '
+                '-e "AK={2}" upgrade:{3}'.format(
+                    hostname, satellite_hostname, ak, client_os))
+            result[hostname] = container_id
+        return result
+    except ValueError:
+        print 'Clients count to generate on Docker should be atleast 1 !'
+
+
+def get_latest_repo_version(cv_name):
```
Comment (jyejare) [inline upgrade_tests/helpers/scenarios_utils.py]: rename it to 'get_latest_cv_version' and move this to automation-tools/satellite6/hammer.py if it not already.

File: upgrade_tests/helpers/scenarios_utils.py [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+    cv_version_data = hammer(
+        'content-view version list --content-view {} '
+        '--organization-id 1'.format(cv_name))
+    latest_cv_ver = sorted([float(data['name'].split(
+        '{} '.format(cv_name))[1]) for data in cv_version_data]).pop()
+    return get_attribute_value(cv_version_data, '{0} {1}'.format(
+        cv_name, latest_cv_ver), 'id')
+
+
+def dockerize(ak_name):
```
Comment (jyejare) [inline upgrade_tests/helpers/scenarios_utils.py]: What is the difference between this and 'generate_satellite_docker_clients_on_rhevm_upgrade' function ? Cant you make it one to suite your need.

File: upgrade_tests/helpers/scenarios_utils.py [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+    return clients
+
+
+def get_satellite_host():
+    """Get the satellite hostname depending on which jenkins variables are set
+
+        :return string : Returns the satellite hostname
+
+        Environment Variable:
+
```
Comment (jyejare) [inline upgrade_tests/helpers/scenarios_utils.py]: Remove this blank line. Use flake8 for your code check.

File: upgrade_tests/test_scenarios.py [hunk 13]
```diff
@@ -0,0 +1,251 @@
+# -*- encoding: utf-8 -*-
+"""Test for Upgrade Scenario's
+
+:Requirement: Satellite upgrade
+
+:CaseAutomation: Automated
+
+:CaseLevel: Acceptance
+
+:CaseComponent: UI
```
Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: Should be CLI ?

File: upgrade_tests/test_scenarios.py [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+"""
+import os
+import time
+
+from automation_tools.satellite6 import hammer
+from upgrade.helpers.docker import (
+    docker_execute_command,
+    refresh_subscriptions_on_docker_clients
+)
+from upgrade_tests.helpers.scenarios_utils import (
```
Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: rename 'scenarios_utils' to 'scenarios'.

File: upgrade_tests/test_scenarios.py [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+    rpm2
+)
+from fabric.api import execute, run
+from fauxfactory import gen_alpha
+from unittest2.case import TestCase
+from upgrade_tests import post_upgrade, pre_upgrade
+
+
+class ScenarioBug1429201(TestCase):
+    """This Class will server as a whole scenario with pre-upgrade and
```
Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: s/server/serve/

Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: Need more explanation specially about bug

File: upgrade_tests/test_scenarios.py [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+    post-upgrade test-case
+    """
+    prd_name = gen_alpha()
```
Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: gen_alpha() generated strings will trouble us during debugging this scenario.

Suppose you have 4 scenarios and you have created 5 products during these scenarios execution, then it would be very difficult to link up that product with the scenario. We should have some trace of scenario with the each entity that we are creating during upgrade.

with that saying, lets change to this:
```
prd_name = 'scenario1-'+gen_alpha()
```
So that debugging will become very easy and we need not dive in all entities to get their links with scenarios.

Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: Also assign a scenario number for each scenario testclass in docstrings.

File: upgrade_tests/test_scenarios.py [hunk 17]
```diff
@@ -0,0 +1,251 @@
...
+    custom_repo = 'https://' + sat_host + '/pub/custom_repo/'
+    _, rpm1_name = os.path.split(rpm1)
+    _, rpm2_name = os.path.split(rpm2)
+
+    def setUp(self):
+        hammer.set_hammer_config()
+
+    @pre_upgrade
+    def test_pre_user_scenario_bug_1429201(self):
+        """This is pre-upgrade scenario test to verify if we can create a
```
Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: You can copy this to test class.

File: upgrade_tests/test_scenarios.py [hunk 18]
```diff
@@ -0,0 +1,251 @@
...
+         """
+        try:
+            run('rm -rf {0}'.format(self.file_path))
+            run('mkdir {0}'.format(self.file_path))
+        except OSError:
+            run('mkdir /var/www/html/pub/custom_repo')
+        run('wget {0} -P {1}'.format(rpm1, self.file_path))
+        run('createrepo --database {0}'.format(self.file_path))
+        # End to End product + ak association
+        print hammer.hammer_product_create(self.prd_name, self.org_id)
```
Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: Why print statements ?

File: upgrade_tests/test_scenarios.py [hunk 19]
```diff
@@ -0,0 +1,251 @@
...
+                                                  self.lc_name
+                                                  )
+        print hammer.hammer_activation_key_add_subscription(self.ak_name,
+                                                            self.org_id,
+                                                            self.prd_name
+                                                            )
+        time.sleep(5)
+        # Creating vm and subscribing to AK
+        container_ids = dockerize(self.ak_name)
+        print container_ids.values()[0]
```
Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: ??

File: upgrade_tests/test_scenarios.py [hunk 20]
```diff
@@ -0,0 +1,251 @@
...
+            container_ids.values()[0],
+            'yum list {0} | grep {0}'.format(self.rpm1_name.split('-')[0]),
+            host=self.docker_vm)
+        # Info on created entities to assert the test case using hammer info
+        prd_info = hammer.hammer('"{0}" info --name "{1}" --organization-id '
+                                 '{2}'.format('product',
+                                              self.prd_name,
+                                              self.org_id
+                                              ))
+        get_name = hammer.get_attribute_value(prd_info, self.prd_name, 'name')
```
Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: rename get_name to something informative.

File: upgrade_tests/test_scenarios.py [hunk 21]
```diff
@@ -0,0 +1,251 @@
...
+            'ak_name': self.ak_name,
+            'repo_name': self.repo_name,
+            'container_ids': container_ids
+        }
+        }
+        create_dict(global_dict)
+
+    @post_upgrade
+    def test_post_user_scenario_bug_1429201(self):
+        """This is post-upgrade scenario test to verify if we can alter the
```
Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: Copy this bit to test class.

File: upgrade_tests/helpers/scenarios_utils.py (hunk 11 above)
Comment (san7ket) [inline upgrade_tests/helpers/scenarios_utils.py]: will use the one in automation , and will remove this, "generate_satellite_docker_clients_on_rhevm_upgrade" .
Dockerize function , will create the Docker_VM is not already up and will call the "generate_satellite_docker_clients_on_rhevm_upgrade"

File: upgrade_tests/helpers/scenarios_utils.py (hunk 10 above)
Comment (san7ket) [inline upgrade_tests/helpers/scenarios_utils.py]: Can we add this as a helper function, In the test class itself , as it will be used by the tests only and act as test helper.

File: upgrade_tests/helpers/scenarios_utils.py (hunk 5 above)
Comment (jyejare) [inline upgrade_tests/helpers/scenarios_utils.py]: We need a separate PR for this to collect all tests entities data in one file. This will overwrite the last test data. And we should append the entity of test case and not write again.

Comment (jyejare) [inline upgrade_tests/helpers/scenarios_utils.py]: As of now I am ok with this implementation as there is only one test case. Going forward we should take care abt this.

File: upgrade_tests/helpers/scenarios_utils.py (hunk 10 above)
Comment (jyejare) [inline upgrade_tests/helpers/scenarios_utils.py]: No add that to hammer, as its specific to hammer. Also tomorrow we will use robotello as our dependency for hammer tasks. And there we will get everything that is required.

File: upgrade_tests/helpers/scenarios.py [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+    logger.info('Generating client on RHEL7 on Docker. '
+                'Please wait .....')
+    # First delete if any containers running
+    execute(
+        remove_all_docker_containers, only_running=False, host=docker_vm)
+    # Generate Clients on RHEL 7
+    time.sleep(30)
+    clients = execute(
+        generate_satellite_docker_clients_on_rhevm,
+        'rhel7',
```
Comment (jyejare) [inline upgrade_tests/helpers/scenarios.py]: Why only rhel7 ?

File: upgrade_tests/helpers/scenarios_utils.py (hunk 11 above)
Comment (jyejare) [inline upgrade_tests/helpers/scenarios_utils.py]: Ok Let it be, Later we will update this if needed, but I am not happy with duplicating code. We should have updated the existing to match this.

File: upgrade_tests/helpers/scenarios_utils.py (hunk 5 above)
Comment (san7ket) [inline upgrade_tests/helpers/scenarios_utils.py]: No, if there are multiple test cases , it all goes into entities_data , it won't overwrite AFAIK , let me check once to confirm.

File: upgrade_tests/test_scenarios.py (hunk 16 above)
Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: I dont see suggested changes in PR.

File: upgrade_tests/test_scenarios.py [hunk 23]
```diff
@@ -0,0 +1,276 @@
...
+    def create_repo(self):
+        try:
+            run('rm -rf {0}'.format(self.file_path))
+            run('mkdir {0}'.format(self.file_path))
+        except OSError:
+            run('mkdir /var/www/html/pub/custom_repo')
+        run('wget {0} -P {1}'.format(rpm1, self.file_path))
+        run('createrepo --database {0}'.format(self.file_path))
+
+    def get_latest_cv_version(self, cv_name):
```
Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: Move this to hammer! Keep as less common functions in test class as possible.

File: upgrade_tests/helpers/scenarios.py (hunk 22 above)
Comment (san7ket) [inline upgrade_tests/helpers/scenarios.py]: Any distro would do for this test , but yea that could be passed as an argument to the dockerize function, which will flexibility to choose distro from test itself.

File: upgrade_tests/test_scenarios.py (hunk 18 above)
Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: This way the stdouts on job run will increase, I am really against of putting so ma ny print statements. 
The stdouts will be huge with increasing test scenarios. Please remove them.

File: upgrade_tests/test_scenarios.py [hunk 24]
```diff
@@ -0,0 +1,276 @@
...
+            container_ids.values()[0],
+            'yum list {0} | grep {0}'.format(self.rpm1_name.split('-')[0]),
+            host=self.docker_vm)
+        # Info on created entities to assert the test case using hammer info
+        prd_info = hammer.hammer('"{0}" info --name "{1}" --organization-id '
+                                 '{2}'.format('product',
+                                              self.prd_name,
+                                              self.org_id
+                                              ))
+        self.assertEqual(self.prd_name,
```
Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: Update this :
```
self.assertEqual(self.prd_name,
                         hammer.get_attribute_value(prd_info,
                                                    self.prd_name,
                                                    'name'))
```
as:
```
self.assertEqual(
    self.prd_name,
    hammer.get_attribute_value(prd_info, self.prd_name,  'name')
)
```

File: upgrade_tests/test_scenarios.py [hunk 25]
```diff
@@ -0,0 +1,276 @@
...
+             5. Create content view CVFoo
+             6. Add RepoFoo to CVFoo
+             7. Publish version 1 of CVFoo
+
+         :expectedresults: The client and product is created successfully
+         """
+        execute(self.create_repo, self, host=self.sat_host)
+        # End to End product + ak association
+        print hammer.hammer_product_create(self.prd_name, self.org_id)
+        print hammer.hammer_repository_create(self.repo_name,
```
Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: Update this style to:
```
hammer.hammer_repository_create(
    self.repo_name,
    self.org_id,
    self.prd_name,
    self.custom_repo
)
```

File: upgrade_tests/test_scenarios.py [hunk 26]
```diff
@@ -0,0 +1,276 @@
...
+                                              self.org_id,
+                                              self.prd_name,
+                                              self.custom_repo
+                                              )
+
+        print hammer.hammer('lifecycle-environment create --name "{0}" '
```
Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: Follow above.

File: upgrade_tests/test_scenarios.py [hunk 27]
```diff
@@ -0,0 +1,276 @@
...
+                            '--organization-id {1} --prior-id "{2}"'.format(
+                                                                self.lc_name,
+                                                                self.org_id,
+                                                                1))
+        print hammer.hammer_repository_synchronize(self.repo_name,
```
Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: Follow above.

File: upgrade_tests/test_scenarios.py [hunk 28]
```diff
@@ -0,0 +1,276 @@
...
+                                                   self.org_id,
+                                                   self.prd_name
+                                                   )
+        print hammer.hammer_content_view_create(self.cv_name, self.org_id)
+        print hammer.hammer_content_view_add_repository(self.cv_name,
```
Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: Follow above.

File: upgrade_tests/test_scenarios.py [hunk 29]
```diff
@@ -0,0 +1,276 @@
...
+                                                        self.org_id,
+                                                        self.prd_name,
+                                                        self.repo_name
+                                                        )
+        print hammer.hammer_content_view_publish(self.cv_name, self.org_id)
+        latest_repo_version = self.get_latest_cv_version(self.cv_name)
+
+        lc_result = hammer.hammer('"{0}" info --name "{1}" --organization-id '
```
Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: Follow above.

File: upgrade_tests/test_scenarios.py [hunk 30]
```diff
@@ -0,0 +1,276 @@
...
+                                  '{2}'.format('lifecycle-environment',
+                                               self.lc_name,
+                                               self.org_id
+                                               ))
+        lifecycle_id = hammer.get_attribute_value(lc_result,
```
Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: Follow above.

File: upgrade_tests/test_scenarios.py [hunk 31]
```diff
@@ -0,0 +1,276 @@
...
+                                                  self.lc_name,
+                                                  'id'
+                                                  )
+        print hammer.hammer_content_view_promote_version(self.cv_name,
```
Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: Follow above.

File: upgrade_tests/test_scenarios.py [hunk 32]
```diff
@@ -0,0 +1,276 @@
...
+                                                         latest_repo_version,
+                                                         lifecycle_id,
+                                                         self.org_id
+                                                         )
+        print hammer.hammer_activation_key_create(self.ak_name,
```
Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: Follow above.

File: upgrade_tests/test_scenarios.py [hunk 33]
```diff
@@ -0,0 +1,276 @@
...
+                                                  self.org_id,
+                                                  self.cv_name,
+                                                  self.lc_name
+                                                  )
+        print hammer.hammer_activation_key_add_subscription(self.ak_name,
```
Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: Follow above.

File: upgrade_tests/test_scenarios.py [hunk 34]
```diff
@@ -0,0 +1,276 @@
...
+        # Creating vm and subscribing to AK
+        container_ids = dockerize(self.ak_name)
+        time.sleep(30)  # Subscription manager needs time to register
+        result = execute(
+            docker_execute_command,
+            container_ids.values()[0],
+            'yum list {0} | grep {0}'.format(self.rpm1_name.split('-')[0]),
+            host=self.docker_vm)
+        # Info on created entities to assert the test case using hammer info
+        prd_info = hammer.hammer('"{0}" info --name "{1}" --organization-id '
```
Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: Follow above.

**Comment by jyejare** [comment]
@san7ket , Please follow flake8 and update your PR accordingly.

File: upgrade_tests/test_scenarios.py [hunk 35]
```diff
@@ -0,0 +1,288 @@
...
+            host=self.docker_vm
+        )  # should be error
+        result_pass = execute(
+            docker_execute_command,
+            container_ids.values()[0],
+            'yum install -y {0}'.format(self.rpm2_name.split('-')[0]),
+            host=self.docker_vm
+        )  # should be successful
+        self.assertEqual(pkcl_ak_name,
+                         hammer.get_attribute_value(
```
Comment (jyejare) [inline upgrade_tests/test_scenarios.py]: Indent this properly.
